### PR TITLE
Add watcher for continuously watching service 

### PR DESCRIPTION
**Pull request by rushi47**
Add informer to keep watching Service Endpoints.
- [x] Handle updates 
- [x] New Service is added based on filter : make sure global service for it exists

### CODE REVIEW

**Comment by rushi47** [comment]
Very minimal indirection. Adds watcher(Informer), and make sure it keeps running.

File: service-mirror.go
Comment (mateiidavid) [inline service-mirror.go:28]: Hmmm this is good, but we should add a flag for it maybe. The services should use the same namespace as the controller. When we run locally, we'd expect the namespace to be `""`. When we deploy, we'd expect it to be the same namespace as the controller. Here's an example:

```
# Run locally, don't pass in a namespace. Default namespace in k8s is also "".
go run ./main.go 
```

```
# A hypothetical deployment template spec:
- name: rushi-controller:v0.0.1
  args:
  - --controller-namespace
  - "global-prototype" # in helm we can template this {{ .Values.namespace }}
 ```

File: service-mirror.go
Comment (mateiidavid) [inline service-mirror.go:30]: Naming nit: we are not getting anything we are creating something. How about `mkWatchFilter` or `createWatcherFilter`? 

However, looking at this, maybe it should ideally be stored in a struct (along with other info we will need, e.g. client)?

```go
struct ServiceWatcher {
   client: k8s.API,
   labelSelector: labels.Selector
}

func (sw *ServiceWatcher) withServiceSelector(l labels.Selector) {
   sw.labelSelector = l
} 

func (sw *ServiceWatcher) WithDefaultSelector(l labels.Selector) {
    mirrorSvcLabel, err := labels.NewRequirement("mirror.linkerd.io/mirrored-service", selection.Equals, []string{"true"})
	if err != nil {
		ctx.log.Errorf("Unable to generate error requirement, Err : %v", err.Error())
	}

	// Create label requirements for "mirror.linkerd.io/headless-mirror-svc-name" not existing
	mirrorSvcParentLabel, err := labels.NewRequirement("mirror.linkerd.io/headless-mirror-svc-name", selection.DoesNotExist, []string{})
	if err != nil {
		ctx.log.Errorf("Unable to generate error requirement, Err : %v", err.Error())
	}

	// Create the label selector
	svcFilter := labels.NewSelector()
	svcFilter = svcFilter.Add(*mirrorSvcParentLabel)
	svcFilter = svcFilter.Add(*mirrorSvcLabel)
  
   sw.withSelector(svcFilter)
}
```

We could also offer a nicer API around this maybe and not make the `withDefaultSelector` public:

```go
func NewDefaultWatcher(client) ServiceWatcher {
   sw := &ServiceWatcher{client}
   sw.withDefaultSelector()
   sw
}

func main()
   client := makeClient()
  sw := sw.NewDefaultWatcher(client)
  sw.Run()
}
```

wdyt?

**Review by mateiidavid — COMMENTED** [review]
Looks good so far. Think we've reached a point where it makes sense to start modelling our watcher a little bit and provide a tiny bit of abstraction. Let's maybe move the watcher stuff in its own file, and keep the main function to handle the process lifecycle.

I'd say
1.  main function: parses args, creates client, creates & runs watcher
2. watcher: collects all information it needs (labels, client, whatever) and provides a run function that we can call in main.

File: service-mirror.go
Comment (rushi47) [reply service-mirror.go:28]: Yess I agree with Flag but i have doubt regarding running controller in same namespace as spawning global-svc. I am smh kind biased towards putting it in default as other services will be mirrored across different namespace. As global service is aggregator will put it in default but i agree that this is should be configurable.

Regarding `""` in using namespace, smh client-go was giving me issue when I was creating service. Error was kind of : if you creating or giving resource name, namespace couldn't be empty

File: service-mirror.go
Comment (rushi47) [reply service-mirror.go:30]: +1 to this. This looks pretty nit, i will refactor my code :) and  as you have also give go ahead, I will do restructuring of code.

Comment (rushi47) [reply service-mirror.go]: Done added all fixes, main function, structs

**Comment by rushi47** [comment]
Done, refactored the logic.

File: service-mirror.go
Comment (mateiidavid) [reply service-mirror.go:17]: Hmm, you're right that the controller should put it in its own namespace. Perhaps we don't need a flag for this at all. What if instead we just adapt to the services we are mirroring?

Let's take two examples:

```
msql service mirrors from target1 and target2

. mysql namespace
|___ mysql (service created by source)
|___ mysql-target1
|___mysql-target2

redis service mirror from target1

.redis namespace
|___ redis
|___redis-target1
```

Let's ignore for a while that someone would use both 😅 . We can see that people would expect the mirror to be in the _actual_ namespace as the service they have. Typically, applications will live in the same namespace too. We should do the same and add the global mirror in whatever namespace the rest of the services are deployed in.

Namespaces typically mean ownership in k8s, and they have a few other implications:
1. calling services is easier (you don't need the whole fqdn, just the service name and port)
2. routing and policy are namespace dependent; dynamic routing and policy for example takes its scope depending on which namespace the routes are declared in.

Typically a "service owner" (for a database it would be the platform team) manage their service and give some routing defaults. Other clients are free to consume the service and provide their own policy on top of it. 

Let's merge this as is now (default namespace) and create an issue for it.

File: justfile
Comment (mateiidavid) [inline justfile:14]: hate to be a stickler for this but can we rename `service-mirror.go` to `main.go`? We should probably set it up with a pkg (or internal better yet) directory too. Let's do it in a follow up don't wanna block on it.

File: service-mirror.go
Comment (mateiidavid) [reply service-mirror.go:72]: Okay, I actually don't think we should support this at all. It adds a bit of complexity (moar indirection!!!). Here's my view on it.

* we maintain a struct definition that wraps around a logger, a context and a client
* we only use this to pass it in the constructor
* service watcher then calls them by going through this composed struct `watcher.ctx.log` (imagine writing that for every single log line).

I would remove it personally. If you want it to stay in and you anticipate it will hold more information, we can embed it instead so the fields are flattened

```go
struct Watcher{
  Context
}

// and now we can do watcher.log
```

We can always bring it back in the future. Remember, it's easier to add something new in than it is to remove something.

File: service.go
Comment (mateiidavid) [inline service.go:21]: Should this be renamed to `Watcher`? What if we start watching endpointslices?

File: service.go
Comment (mateiidavid) [inline service.go:25]: hmmm is this comment accurate? do we install the service watcher or mirror the services in there?

File: service.go
Comment (mateiidavid) [inline service.go:40]:
```suggestion
func (svc *ServiceWatcher) checkNamespaceExists() {
```

maybe redundant to say if 🤷🏻

File: service.go
Comment (mateiidavid) [inline service.go:64]: This is good, but what if the error is actually not nil here? We'll still create a namespace. However, if our error is related to rbac, or a transient network failure, socket's been dropped, etc (you get it), then we'll make it harder on ourselves to actually isolate the root issue.

```suggestion
    if err != nil {
        if apiError.IsAlreadyExists(err) {
		      log.Debugf("Skipped creating namespace '%s'; already exists", namespace)
		      return
		}
		log.Errorf("Failed to get namespace '%s': %v", err)
		return
	} 
     // Create the namespace
	newNamespace := &corev1.Namespace{
			ObjectMeta: metav1.ObjectMeta{
				Name: namespace,
			},
		}
		_, err := client.CoreV1().Namespaces().Create(svc.ctx.ctx, newNamespace, metav1.CreateOptions{})
		if err != nil {
			log.Errorf("Issue creating namespace : %v", err.Error())
		}

		log.Infof("Namespace '%s' created", namespace)
	}
}
```

It's also generally better to return early instead of nesting in ifs

File: service.go
Comment (mateiidavid) [inline service.go:69]: I'd avoid pinning here, see my suggestion above about embedding.

File: service.go
Comment (mateiidavid) [inline service.go:65]: Great, this looks good!

**Review by mateiidavid — APPROVED** [review]
I think the rest of this we can work out as we go, for now it looks good except the things I commented on. Would be good to fix before merging. Great job so far!

File: service-mirror.go
Comment (rushi47) [reply service-mirror.go:17]: Thanks for the explanation matei. Well to add to your last point so i have already added flag if its not provided its default to default 😅 . Else it uses the namespace provided. But with your first point that make sense. Probably later we can make both cases are handled.

File: service-mirror.go
Comment (rushi47) [reply service-mirror.go:72]: Yess this make sense let me refactor it and try to do in suggested way. Its odd rn.

File: service.go
Comment (rushi47) [reply service.go:21]: So in EndpointSlices, I have done this : Its kind of parent struct 
https://github.com/rushi47/service-mirror-prototype/pull/10/files#diff-0ea94270ccbc3fa46cbb5b0c86ea055db059190c02ca103a32426192f6ca3d78R27-R33

Which is extended by EndpointWatcher and ServiceWatcher
svc watcher : https://github.com/rushi47/service-mirror-prototype/pull/10/files#diff-c52daedaeda258db15ec11d57bbf299412a20328f6cd026980ef07f01960dd3aR22-R24

ep watcher : https://github.com/rushi47/service-mirror-prototype/pull/10/files#diff-92d7790692abf52c09d20d29b7945730bed18aac6643dbc4488a2b530e9390d7R20-R24

ps : This solves kind of problem of logging but not in straightforward way. I will change with what you suggested above. But to main your main question i hope this helps.

File: service.go
Comment (rushi47) [reply service.go:25]: Oops yep ! mirror it.

File: service.go
Comment (rushi47) [reply service.go:40]: +1

File: service-mirror.go
Comment (rushi47) [reply service-mirror.go:97]: So `channel` kind of does the trick. First we create an `unbuffered ` channel and then defer to close it. As no struct is sending value over this channel it waits for only SIGTERM. So it runs indefinitely till we send the signal. Resulting we can avoid using select.

File: service-mirror.go
Comment (mateiidavid) [inline service-mirror.go:91]: Should we sync before we run?

File: service-mirror.go
Comment (mateiidavid) [inline service-mirror.go:22]: I would do a field name on these and then embed in the watcher so we can do watcher.client or watcher.ctx

File: service.go
Comment (mateiidavid) [inline service.go:1]: What are we going to do with this file?

**Review by mateiidavid — APPROVED** [review]

File: service-mirror.go
Comment (rushi47) [reply service-mirror.go:91]: I derived this from one of example : https://www.cncf.io/blog/2019/10/15/extend-kubernetes-via-a-shared-informer/

From my understanding, the way this works is because of shared channel. In waitforcachesync we use the same channel, which we pass while running informer. So  waitforcachesync can block the go routine. Till the informer hasn't synced the cache.

File: service-mirror.go
Comment (rushi47) [reply service-mirror.go:22]: This is still part of context struct, it had named previously so it will be again like svc.ctx.client or svc.ctx.ctx.

Comment (mateiidavid) [reply service-mirror.go:22]: If you embed you can refer to the field directly. If your watcher holds an embdded context you can do watcher.client

File: service.go
Comment (rushi47) [reply service.go]: Done updated added main.go. and fixes structured.

**Pull request merged**
